enhancements/microshift/periodic-etcd-defragmentation-in-microshift.md

```diff
@@ -0,0 +1,164 @@
+
+There are a few issues with this approach:
+1. It relies on cron being installed on the system and running.
+2. If Microshift is not running, the cronjob might still fire and would fail because it would be able to reach etcd which would cause a lot of false-positive failures to appear in the log.
+3. If we wanted to cover both approaches (pure periodic and threshold), then we would likely need two different cronjobs added to the tab during install.
+
+#### Defragmentation from transient systemd unit
+This would be similar to the work to launch etcd itself as a transient systemd unit. A new sub-command could be added to Microshift (or microshift-etcd) that would do much the same as the goroutine approach above: two timeouts in a loop (assuming doing both pure periodic and threshold) that will launch the defragmentation.
+Its runtime would be attached to that of Microshift and, like the goroutine, it would be easy to detect when/if a defragmentation was running or when/if it failed.
+
+This approach would be the most like an operator in OCP.
+
+### Workflow Description
+For the end user, any change we make should be completely transparent. Defragmentation is only a problem when it isn't set up (or it fails frequently) and etcd runs out of resources, causing a disruption.
+For a Microshift Administrator, they should be able to easily know when a defragmentation is occuring/occured and if it has failed.
+
+#### Variation [optional]
+TODO
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+TODO
+
+### Risks and Mitigations
+
+Some methods - cron-job and systemd unit - could potentially fail silently (or at least force monitoring of a non-Microshift location: cron logs/journalctl) and lead the defragmentation to not execute, which could allow etcd to grow and hit its quota which would cause further writes to fail.
+The gorountine might be superior in this case as it could send logs to the same stream as normal Microshift, which would be more likely noticed by an Administrator.
+To mitigate the above issue, we could have the systemd unit's logs streamed into the Microshift log stream; it would essentially be a subprocess of Microshift.
+
+### Drawbacks
+
+Regardless of which method is chosen, during the duration of the defragmentation, etcd will be unavailable for writes to the database; clients may still read data, but writes will be denied until the defragmentation is finished. Given the size of Microshift, this hold should not be very long, likely at most single seconds; this would be a good thing to test to ensure the disruption duration is acceptable.
+
+
+## Design Details
+
+### Open Questions [optional]
+
+- Do we even need/want automatic defrag?
+  - Since these are small clusters, is the short hold on writes (while the defrag takes place) acceptable for this application.
```

Comment:
We could, but there's still a chance that during run, especially longer running clusters, that we need to defrag inflight otherwise etcd would run out of space.

Reply:
I've added this an alternative to pure periodic defrag.